### ISSUE TITLE
"Unknown cursor type" with vmware 6.5 

### ISSUE THREAD

**Issue by sthen** [TigerVNC/tigervnc#1409]
**Describe the bug**
If I connect to the native VNC server on VMware 6.5.0U3 and move the cursor over the window, tigervnc's vncviewer throws an "Unknown cursor type" exception. If I modify CMsgReader.cxx to display the type, it is 255. In almost every occasion this happens right after connect (there have been some rare occasions where it doesn't trigger then, but very few - probably triggers on 95%+ of attempts or more).

It's working in 1.11.0. I've built from multiple commits between there and 1.12.0; I haven't hit it in 6d1603c3da but do in 59d5bc4095 ("merge branch noblock").

**To Reproduce**
vncviewer $vmware_host:port
enter password
move cursor over window

**Expected behavior**
connection to remain working

**Screenshots**
```
$ vncviewer.tigervnc  10.15.5.15:1 

TigerVNC Viewer 64-bit v1.12.0
Built on: 2022-01-13 18:04
Copyright (C) 1999-2021 TigerVNC Team and many others (see README.rst)
See https://www.tigervnc.org for information on TigerVNC.

Thu Jan 13 18:44:40 2022
 DecodeManager: Detected 4 CPU core(s)
 DecodeManager: Creating 4 decoder thread(s)
 CConn:       Connected to host 10.15.5.15 port 5901
 CConnection: Server supports RFB protocol version 3.8
 CConnection: Using RFB protocol version 3.8
 CConnection: Choosing security type VncAuth(2)

Thu Jan 13 18:44:42 2022
 CConn:       Using pixel format depth 24 (32bpp) little-endian rgb888

Thu Jan 13 18:44:46 2022
 CConn:       Unknown cursor type
```

**Client (please complete the following information):**
 - OS: OpenBSD -current
 - VNC client: TigerVNC
 - VNC client version: 1.12.0
 - Client downloaded from: git archive

**Server (please complete the following information):**
 - OS: VMware ESXi 6.5.0U3
 - VNC server: vmware internal
 - VNC server version: vmware internal
 - Server downloaded from: vmware internal
 - Server was started using: RemoteDisplay.vnc.enabled = "TRUE" / RemoteDisplay.vnc.port = "5901" / RemoteDisplay.vnc.key = "xxx"

**Additional context**
I'm the port maintainer on OpenBSD and trying to update the port, however one of the machines I need to connect to on a regular basis is affected by this. Please let me know if there's more information that I can provide! Thanks.

**Comment by CendioOssman** [comment]
Do you think you could run `git bisect` and narrow things down even further? These kinds of bugs are difficult to debug, so having the exact commit that breaks things would be very useful.

**Comment by sthen** [comment]
There didn't seem much point in testing the commits between 6d1603c and 59d5bc4 as they didn't change the source code?

**Comment by sthen** [comment]
Ah, or did you mean on the branch?

**Comment by CendioOssman** [comment]
Yes. There is a high risk that it's still the big commit (ad0f0618fa2ca13d7b916f22eccc5ba3201482cb). But if we're lucky there is some smaller change that is causing this.

**Comment by CendioOssman** [comment]
And `git bisect` is generally better to use than trying to figure out relevant commits on your own. It is clever enough to handle branches and merges.

**Comment by sthen** [comment]
`git bisect` doesn't really work for the way I'm building (I'll have to figure out how to manually replicate the cmake setup done automatically by the ports setup, apply patches because we don't have PAM, etc) so doing it by hand is mostly simpler, anyway I've tried ad0f061 where the problem exists, and the parent commit c0dac22 where it doesn't, so sadly it does seem to be in the big commit.

**Comment by CendioOssman** [comment]
That's a shame. :/

Do you think it is possible to get a traffic capture of things going wrong? If it happens right after connect it should hopefully not be very large.

**Comment by sthen** [comment]
With ad0f0618 it seems to be taking a little longer to trigger on most (but not all) occasions than it was in 1.12.0 (usually needing a bit of clicking around; VM guest is running Windows and double-clicking to open a window has a higher chance of hitting it). Not 100% sure but possibly the "Heavily reduce window for bandwidth estimate" commit makes it happen more quickly.

Is the capture in https://junkpile.org/vnc3.pcap.gz any help? (started after login and before the error was triggered, it's very small, ~20 packets).

**Comment by CendioOssman** [comment]
Thanks. I can see the error now. Could you test this patch:

```diff
diff --git a/common/rfb/CMsgReader.cxx b/common/rfb/CMsgReader.cxx
index 88a31eca..a30a7759 100644
--- a/common/rfb/CMsgReader.cxx
+++ b/common/rfb/CMsgReader.cxx
@@ -636,11 +636,11 @@ bool CMsgReader::readSetVMwareCursor(int width, int height, const Point& hotspot
   if (!is->hasData(1 + 1))
     return false;
 
+  is->setRestorePoint();
+
   type = is->readU8();
   is->skip(1);
 
-  is->setRestorePoint();
-
   if (type == 0) {
     int len = width * height * (handler->server.pf().bpp/8);
     rdr::U8Array andMask(len);
```

**Comment by sthen** [comment]
Great, that's working. Many thanks!

**Comment by CendioOssman** [comment]
Thanks for testing. Fixed in 1a729dc0129052a028eae4582a6d6d10564d13c4.
